### ISSUE TITLE
Question: How to store an arbitrary value in state?

### ISSUE THREAD

**Issue by gtmtech** [hashicorp/terraform#32270]
### Terraform Version

```shell
N/A
```

### Use Cases

I want to store an arbitrary value in state. I will know the value (passed in by an environment variable), and I just want to store it in state. I imagine there is a provider for this _somewhere_ but I couldn't see an easy one to use .

My actual usecase is that I want to know from which git commit a terraform infra was last deployed. I have all my terraform resources in a module (module.root), so i figure outside the top level module I'll have some arbirary resource like this:

```
resource "some_resource" "git_commit_hash" {
    value = var.git_commit_hash
    depends_on = [ 
        module.root
    ]
}
```

This will mean the some_resource only gets applied when everything in the module.root is successfully applied - great!

By looking at all my git_commit_hash resources across all my statefiles I will understand the last successful githash that was applied. 

### Attempted Solutions

I am looking for a provider or terraform resource that can just store a value, and I cant think of one, despite searching. I'm sure there is one

**Comment by bflad** [comment]
Related: https://github.com/hashicorp/terraform/pull/31757

**Comment by jbardin** [comment]
Hi @gtmtech,

The objects which are stored in state consist of managed resources, and root modules output. The canonical utility resource for storing arbitrary values is the `null_resource`, but at @bflad mentioned, there is a proposal for a more flexible builtin resources to take its place.

**Comment by gtmtech** [comment]
@jbardin I didn't think `null_resource` stored any values in state? - what am I missing?

**Comment by jonwtech** [comment]
How about:
```
resource "local_file" "foo" {
  content  = var.git_commit_hash
  filename = "/dev/null"
  depends_on = [ 
        module.root
  ]
}
```
Not perfect as Terraform will always want to create the resource, but the value is stored in state.

**Comment by jbardin** [comment]
@gtmtech, `null_resource` has a `triggers` attribute which is a map that can store any number of string values. The above PR for a `terraform_data` resource has been merged for v1.4, which has dynamically typed attributes that can store any type.
